src/stash/src/transaction.rs

```diff
@@ -0,0 +1,641 @@
+                                                "unexpected lower, got {:?}, expected {:?}",
+                                                current_upper, lower1
+                                            )));
+                                        }
+
+                                        self.seal(collection_id, &upper1, Some(lower)).await
+                                    },
+                                    async move {
+                                        self.update(collection_id, &entries, Some(lower2)).await
+                                    },
+                                    async move {
+                                        self.compact(collection_id, &compact, Some(upper)).await
+                                    },
+                                )
+                                .await?;
+                            Ok::<_, StashError>((collection_id, self.since(collection_id).await?))
+                        },
+                    );
+                    try_join_all(futures).await
+                })
+            })
+            .await?;
+
+        for cons in consolidations {
+            self.consolidations.send(cons).unwrap();
+        }
+        Ok(())
+    }
+
+    /// Like update, but starts a savepoint.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn update_savepoint(
+        &self,
+        collection_id: Id,
+        entries: &[((Value, Value), Timestamp, Diff)],
+        upper: Option<Antichain<Timestamp>>,
+    ) -> Result<(), StashError> {
+        self.in_savepoint(|| Box::pin(async { self.update(collection_id, entries, upper).await }))
+            .await
+    }
+
+    /// Directly add k, v, ts, diff tuples to a collection.`upper` can be `Some`
+    /// if the collection's upper is already known. Caller must have already
+    /// called in_savepoint.
+    ///
+    /// This function should not be called outside of the stash crate since it
+    /// allows for arbitrary bytes, non-unit diffs in collections, and doesn't
+    /// support transaction safety. Use `TypedCollection`'s methods instead.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn update(
+        &self,
+        collection_id: Id,
+        entries: &[((Value, Value), Timestamp, Diff)],
+        upper: Option<Antichain<Timestamp>>,
+    ) -> Result<(), StashError> {
+        {
+            // Panic if the caller didn't initiate the savepoint.
+            let savepoint = self.savepoint.lock().unwrap();
+            assert!(*savepoint);
+        }
+
+        let mut futures = Vec::with_capacity(entries.len());
+        for ((key, value), time, diff) in entries {
+            futures.push(async move {
+                self.client
+                    .execute(
+                        self.stmts.update_many(),
+                        &[&collection_id, &key, &value, &time, &diff],
+                    )
```

Comment:
I wrote up some code that did this. It's not free because the current version uses a prepared statement, and that would require generating a unique string, preparing it, then executing, so it's not clear where the performance tradeoff there is. Agree it's worthwhile to benchmark in the future, but mostly I suspect we aren't changing enough numbers of rows in the same collection it matters that much?